### PR TITLE
[AIRFLOW-2451] do not insert extra / char when using wildcard

### PR DESCRIPTION
**Pull request by berislavlopac**
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the bug reported at [AIRFLOW-2451](https://issues.apache.org/jira/browse/AIRFLOW-2451). 

### Description
- [x] When using `destination_object` while having a wildcard in `source_object`, the value of the former is prepended to the actual path the file is copied to. The original code is adding a `/` character after the prefix, assuming that the goal is to specify a "folder" which is not necessarily the case; that behaviour also results in `//` when prefix already includes a slash, which in GCS is not normalised to a single slash as is the case in POSIX. This PR removes this extra character and adds some optimisation to the code, as well as additional documentation and tests.

### Tests
- [x] The included tests cover several scenarios when using wildcard in combination with `description_object`

### Documentation
- [x] Example explaining the behaviour added to the docstring.

### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: airflow/contrib/operators/gcs_to_gcs.py [hunk 1]
```diff
@@ -73,40 +79,39 @@ def __init__(self,
...
         )
+        log_message = 'Executing copy of gs://{0}/{1} to gs://{2}/{3}'
 
-        if '*' in self.source_object:
-            wildcard_position = self.source_object.index('*')
-            objects = hook.list(self.source_bucket,
-                                prefix=self.source_object[:wildcard_position],
-                                delimiter=self.source_object[wildcard_position + 1:])
+        if self.wildcard in self.source_object:
+            prefix, delimiter = self.source_object.split(self.wildcard, maxsplit=1)
```
Comment (kaxil) [inline airflow/contrib/operators/gcs_to_gcs.py]: The parameter `maxsplit` is not available in python 2. Hence the tests are failing for Py2. 


```
20) ERROR: test_execute_wildcard_with_destination_object_retained_prefix (tests.contrib.operators.test_gcs_to_gcs_operator.GoogleCloudStorageToCloudStorageOperatorTest)
----------------------------------------------------------------------
   Traceback (most recent call last):
    .tox/py27-backend_postgres/lib/python2.7/site-packages/mock/mock.py line 1305 in patched
      return func(*args, **keywargs)
    tests/contrib/operators/test_gcs_to_gcs_operator.py line 127 in test_execute_wildcard_with_destination_object_retained_prefix
      operator.execute(None)
    airflow/contrib/operators/gcs_to_gcs.py line 102 in execute
      prefix, delimiter = self.source_object.split(self.wildcard, maxsplit=1)
   TypeError: split() takes no keyword arguments
```

Change this to:
```
prefix, delimiter = self.source_object.split(self.wildcard, 1)
```

This should fix the errors

**Review by kaxil — CHANGES_REQUESTED** [review]

**Comment by berislavlopac** [comment]
Duh of course, been working with 3.6 so much lately that I've completely forgot to run the tests for 2.7... :roll_eyes: Thanks!

**Comment by kaxil** [comment]
I will test this one today. If it works fine I am happy to merge it.

**Comment by kaxil** [comment]
The removal of trailing slash is giving incorrect results:

I tried to run below code to test this

```python
move_gcs = GoogleCloudStorageToGoogleCloudStorageOperator(
        task_id='move_gcs',
        source_bucket='bucket1',
        source_object='example_3/avro/*.avro',
        destination_bucket='bucket1',
        destination_object='example_3_copy',
        google_cloud_storage_conn_id='airflow-service-account'
    )
```

Bucket1 contains following files:
```
gs://bucket1/example_3/avro/usa_names_1910_current_000000000000.avro
gs://bucket1/example_3/avro/usa_names_1910_current_000000000001.avro
gs://bucket1/example_3/avro/usa_names_1910_current_000000000002.avro
gs://bucket1/example_3/avro/usa_names_1910_current_000000000003.avro
gs://bucket1/example_3/avro/usa_names_1910_current_000000000004.avro
gs://bucket1/example_3/avro/usa_names_1910_current_000000000005.avro
```

But when I run it using the changes you made the result is as below:
```
gs://bucket1/example_3_copyusa_names_1910_current_000000000000.avro
gs://bucket1/example_3_copyusa_names_1910_current_000000000001.avro
gs://bucket1/example_3_copyusa_names_1910_current_000000000002.avro
gs://bucket1/example_3_copyusa_names_1910_current_000000000003.avro
gs://bucket1/example_3_copyusa_names_1910_current_000000000004.avro
gs://bucket1/example_3_copyusa_names_1910_current_000000000005.avro
```

I want the files to be copied to `example_3_copy` directory, which should result to something like
```
gs://bucket1/example_3_copy/usa_names_1910_current_000000000000.avro
gs://bucket1/example_3_copy/usa_names_1910_current_000000000001.avro
gs://bucket1/example_3_copy/usa_names_1910_current_000000000002.avro
gs://bucket1/example_3_copy/usa_names_1910_current_000000000003.avro
gs://bucket1/example_3_copy/usa_names_1910_current_000000000004.avro
gs://bucket1/example_3_copy/usa_names_1910_current_000000000005.avro
```

**Comment by berislavlopac** [comment]
This is the correct behaviour. Since you haven't given the slash in your `destination_object`, it is not being added in the final destination.

It is important to keep in mind that in GCS, just like in S3, objects are not actually put into "directories" -- object names are simple strings which are for convenience treated as directories if their names include slashes. Therefore, constructing an object's name should not assume that the goal is to add a "directory"; a slash is just another character in the name.

To take your example, if your `destination_object` included a trailing slash -- `example_3_copy/` -- the original behaviour would produce e.g. `gs://bucket1/example_3_copy//usa_names_1910_current_000000000000.avro`; and in GCS two slashes do not get conflated to one as in a filesystem. Instead, your GCS console would show an extra level of "directory" with an empty string for the name.

Not presuming the use of "directories" also makes things more flexible -- it is easier to add a wanted slash in `destination_object` than to remove an unwanted one added by the operator.

Basically, using `destination_object` with the wildcard replaces the prefix (the part before the wildcard) with the `destination_object`; if the former ends with `/` so should the latter.

**Comment by kaxil** [comment]
I agree with you on some aspects. But if you take an example of how `gsutil` works:

```bash
gsutil cp -r 'gs://bucket1/example_3/avro/*.avro' 'gs://bucket1/example_3_copy'
```

will copy all the avro files to `example_3_copy` directory.

But I agree: directory is just an object with trailing `/` for e.g `example_3_copy/` so can you please add some documentation regarding the same as well? i.e. "If you want to copy your source objects to a directory add a trailing `/` ". Also, please add an example usage in the documentation similiar to https://github.com/apache/incubator-airflow/blob/92363490b725cca345298a9f10613257a7500c9a/airflow/contrib/operators/gcs_list_operator.py#L52:L58 or something like the below example and explaining it.

```python
move_gcs = GoogleCloudStorageToGoogleCloudStorageOperator(
        task_id='move_gcs',
        source_bucket='bucket1',
        source_object='example_3/avro/*.avro',
        destination_bucket='bucket1',
        destination_object='example_3_copy/', # Add "/" if destination object needs to be a directory
        google_cloud_storage_conn_id='airflow-service-account'
    )
```

Thanks

**Comment by kaxil** [comment]
https://github.com/apache/incubator-airflow/blob/master/airflow/contrib/operators/gcs_list_operator.py#L52:L58

**Comment by berislavlopac** [comment]
Sure, will add an example (there is one in the docstring, but is much more limited).

Regarding the `gsutil`, it was intentionally designed to mimic the way POSIX commands work, so it also presents the "directory" abstraction. :slightly_smiling_face:

**Comment by kaxil** [comment]
Thanks @berislavlopac . Appreciate the changes.

👍 LGTM
